### PR TITLE
Latest runtime updates break catch all proxy routes when pointing to …

### PR DESCRIPTION
**Pull request by safihamid**
…http trigger functions in the same function app.

### CODE REVIEW

File: src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs [hunk 1]
```diff
@@ -22,19 +22,87 @@ public class ProxyFunctionExecutor : IFuncExecutor
...
             _webHookReceiverManager = webHookReceiverManager;
             _secretManager = secretManager;
+
+            _httpPrefix = "api/";
+
+            if (_scriptHostManager.Instance != null)
+            {
+                var json = _scriptHostManager.Instance.ScriptConfig.HostConfig.HostConfigMetadata;
+
+                if (json != null && json["http"] != null && json["http"]["routePrefix"] != null)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: What about VS authored precompiled functions? In such cases, will the route be present in function.json?

File: src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs [hunk 2]
```diff
@@ -22,19 +22,87 @@ public class ProxyFunctionExecutor : IFuncExecutor
...
+                path = path.Remove(0, _httpPrefix.Length);
+            }
+
+            // This is a call to the local function app, before handing the route to asp.net to pick the FunctionDescriptor the following will be run:
+            // 1. If the request maps to a local http trigger function name then that function will be picked.
+            // 2. Else if the request maps to a custom route of a local http trigger function then that function will be picked
+            // 3. Otherwise the request will be given to asp.net to pick the appropriate route.
+            foreach (var func in _scriptHostManager.HttpFunctions.Values)
+            {
+                if (!func.Metadata.IsProxy)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: I thought this proxy executor is only used for proxies? When would this condition be true?

File: src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs [hunk 3]
```diff
@@ -22,19 +22,87 @@ public class ProxyFunctionExecutor : IFuncExecutor
...
+                            if (binding.Metadata.IsTrigger)
+                            {
+                                string functionRoute = null;
+                                var jsonContent = binding.Metadata.Raw;
+                                if (jsonContent != null && jsonContent["route"] != null)
+                                {
+                                    functionRoute = jsonContent["route"].ToString();
+                                }
+
+                                if (!string.IsNullOrEmpty(functionRoute) && path.Equals(functionRoute, StringComparison.OrdinalIgnoreCase))
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: Routes can contain regular expressions and other constraints. How can a string comparison like this work?

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 4]
```diff
@@ -1001,8 +1002,8 @@ private Collection<FunctionMetadata> LoadProxyRoutes(string proxiesJson)
             {
                 try
                 {
-                    // Proxy names should follow the same naming restrictions as in function names.
-                    ValidateName(route.Name, true);
+                    // Proxy names should follow the same naming restrictions as in function names. If not, invalid characters will be removed.
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: This doesn't seem related to the hotfix - can you explain this?

File: src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs [hunk 5]
```diff
@@ -22,19 +22,87 @@ public class ProxyFunctionExecutor : IFuncExecutor
...
+
         private WebHookReceiverManager _webHookReceiverManager;
 
         internal ProxyFunctionExecutor(WebScriptHostManager scriptHostManager, WebHookReceiverManager webHookReceiverManager, ISecretManager secretManager)
         {
             _scriptHostManager = scriptHostManager;
             _webHookReceiverManager = webHookReceiverManager;
             _secretManager = secretManager;
+
+            _httpPrefix = "api/";
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: Can you use HttpExtensionConstants.DefaultRoutePrefix here?

File: src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs [hunk 6]
```diff
@@ -22,19 +22,87 @@ public class ProxyFunctionExecutor : IFuncExecutor
...
+
+            if (_scriptHostManager.Instance != null)
+            {
+                var json = _scriptHostManager.Instance.ScriptConfig.HostConfig.HostConfigMetadata;
+
+                if (json != null && json["http"] != null && json["http"]["routePrefix"] != null)
+                {
+                    _httpPrefix = json["http"]["routePrefix"].ToString() + "/";
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: What if the route prefix already has a "/" at the end of it?

File: test/WebJobs.Script.Tests.E2E/ProxyEndToEndTests.cs [hunk 7]
```diff
@@ -71,6 +71,48 @@ public ProxyEndToEndTests(FunctionAppFixture fixture)
 
         [Fact]
         [TestTrace]
+        public async Task LocalFunctionCall2()
```
Comment (mathewc) [inline test/WebJobs.Script.Tests.E2E/ProxyEndToEndTests.cs]: Can you give this a better name than "2"? What makes it different than "1"?

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 8]
```diff
@@ -1106,6 +1107,11 @@ internal static void ValidateName(string name, bool isProxy = false)
             }
         }
 
+        internal static string UpdateProxyName(string name)
+        {
+            return ProxyNameValidationRegex.Replace(name, string.Empty);
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: What is this doing - replacing what with what?

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 9]
```diff
@@ -1106,6 +1107,11 @@ internal static void ValidateName(string name, bool isProxy = false)
             }
         }
 
+        internal static string UpdateProxyName(string name)
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: nit: NormalizeProxyName?

File: src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs (hunk 2 above)
Comment (safihamid) [reply src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: This is our callback logic when the backend is a local function, we go thru the list of FunctionDescriptors, but we don't want this logic for proxy function descriptors. there is some comment above describing the logic.

---
In reply to: [146595677](https://github.com/Azure/azure-webjobs-sdk-script/pull/2060#discussion_r146595677) [](ancestors = 146595677)

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 4 above)
Comment (safihamid) [reply src/WebJobs.Script/Host/ScriptHost.cs]: This is part2 of the hotfix, I sent an email about it last night. many customers are hitting it as well.

---
In reply to: [146596874](https://github.com/Azure/azure-webjobs-sdk-script/pull/2060#discussion_r146596874) [](ancestors = 146596874)

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 8 above)
Comment (safihamid) [reply src/WebJobs.Script/Host/ScriptHost.cs]: This is running the regex Regex(@"[^a-zA-Z0-9_-]", RegexOptions.Compiled | RegexOptions.IgnoreCase);, basically removing all non alphanumeric characters and anything other than '_', and '_'

---
In reply to: [146605862](https://github.com/Azure/azure-webjobs-sdk-script/pull/2060#discussion_r146605862) [](ancestors = 146605862)

File: src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs (hunk 3 above)
Comment (safihamid) [reply src/WebJobs.Script.WebHost/ProxyFunctionExecutor.cs]: Yes this only work for static routes. I can open a separate issue for dynamic routes and investigate later.

---
In reply to: [146596316](https://github.com/Azure/azure-webjobs-sdk-script/pull/2060#discussion_r146596316) [](ancestors = 146596316)

**Comment by DanielSundberg** [comment]
Sorry for asking but I'm confused. Merged to 1.x, what does this mean? Should the fix be in latest 1.0.x? I don't see this mentioned in the change log.

**Comment by safihamid** [comment]
@DanielSundberg Yes the fix is in Functions V1, v1.x is the branch name for V1.
